File: Editorial/1858_Longest_Word_With_All_Prefixes/Java/1_Hash_Set.java
class Solution {

    public String longestWord(String[] words) {
        // Sort the words lexicographically
        Arrays.sort(words);

        // Set to store valid words
        Set<String> validWords = new HashSet<>();
        String longestValidWord = "";

        // Iterate through each word
        for (String currentWord : words) {
            // Check if the word is of length 1 or if its prefix exists in the set
            if (
                currentWord.length() == 1 ||
                validWords.contains(
                    currentWord.substring(0, currentWord.length() - 1)
                )
            ) {
                // Add the current word to the set of valid words
                validWords.add(currentWord);

                // Update the longest valid word if necessary
                if (longestValidWord.length() < currentWord.length()) {
                    longestValidWord = currentWord;
                }
            }
        }

        // Return the longest valid word found
        return longestValidWord;
    }
}

File: Editorial/1801-1900/1858_Longest_Word_With_All_Prefixes/CPP/1_Hash_Set.cpp
class Solution {
public:
    string longestWord(vector<string>& words) {
        // Sort the words lexicographically
        sort(words.begin(), words.end());

        // Set to store valid words
        unordered_set<string> validWords;
        string longestValidWord = "";

        // Iterate through each word
        for (string& currentWord : words) {
            // Check if the word is of length 1 or if its prefix exists in the
            // set
            if (currentWord.length() == 1 ||
                validWords.find(currentWord.substr(
                    0, currentWord.length() - 1)) != validWords.end()) {
                validWords.insert(currentWord);

                if (longestValidWord.length() < currentWord.length()) {
                    longestValidWord = currentWord;
                }
            }
        }

        return longestValidWord;
    }
};
